### PR TITLE
Adding Agent KeyManager V1 facade

### PR DESCRIPTION
**Pull request by azdagron**
The keymanager facade accomodates two "interfaces", the multi (provided by v1) or the single key manager (provided by v0). There is a SVID key manager facade on top the keymanager facade that provides a unified interface strictly for SVID key management. Agent code has been modified to work against this SVID key manager facade.

To reduce some of the friction of introducing this change, the manager is no longer in charge of calling "SetKey" before writing the SVID to disk. Instead, the attestor or rotator packages call "SetKey" after a successful attestation or rotation event before recording the SVID state. This does not change the recovery situation for v0 plugins. However for v1 plugins, this will allow for graceful recovery if writing the SVID fails, since on startup, the agent enumerates all keys stored by the plugin and matches it up to the stored SVID. 

The v1 facade, keymanager/base, keymanager/test, memory, and disk plugins are more or less copied from the server keymanager packages.

The difference in the disk plugin implementation is:
- The configurable is kept consistent (i.e. `directory` instead of `keys_path`). This provides us with a safe upgrade/downgrade path for 1.0. We can change this for 1.1 if we want.
- We store keys in a `keys.json` file within this directory, formatted the same way as the server keymanager plugin of the same name. This gives us a path to migrate the configuration at a later point.
- The `svid.key` is loaded into the key manager entries, and is always updated to the last generated key. This isn't perfect, in that if there is a failed rotation event followed immediately by a downgrade, the key will be incorrect and will cause a re-attestation, but otherwise this should allow for seamless downgrade. 

Many of the changes to existing tests are to accommodate the fact that the SVID key must be managed by a keymanager for things to work properly.

### CODE REVIEW

File: pkg/agent/plugin/keymanager/base/keymanagerbase.go [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+	if req.KeyId == "" {
+		return nil, status.Error(codes.InvalidArgument, "key id is required")
+	}
+	if req.KeyType == keymanagerv1.KeyType_UNSPECIFIED_KEY_TYPE {
+		return nil, status.Error(codes.InvalidArgument, "key type is required")
+	}
+
+	newEntry, err := m.generateKeyEntry(req.KeyId, req.KeyType)
+	if err != nil {
+		return nil, err
```
Comment (amartinezfayo) [inline pkg/agent/plugin/keymanager/base/keymanagerbase.go:93]: I noticed that there are several places where a similar thing happens, including the server side. I'm sorry I missed this in previous reviews.
I think that anything we can do to have errors that help to identify and fix issues is a good thing. I'm curious to know your thoughts.

**Review by amartinezfayo — COMMENTED** [review]
Thanks @azdagron, this is looking great!

File: pkg/agent/manager/manager_test.go [hunk 2]
```diff
@@ -1198,19 +1184,24 @@ func createCA(t *testing.T, clk clock.Clock) (*x509.Certificate, *ecdsa.PrivateK
 	return ca, cakey
 }
 
-func createSVID(t *testing.T, clk clock.Clock, ca *x509.Certificate, cakey *ecdsa.PrivateKey, spiffeID spiffeid.ID, ttl time.Duration) ([]*x509.Certificate, *ecdsa.PrivateKey) {
+func createSVID(t *testing.T, km keymanager.KeyManager, clk clock.Clock, ca *x509.Certificate, cakey *ecdsa.PrivateKey, spiffeID spiffeid.ID, ttl time.Duration) ([]*x509.Certificate, keymanager.Key) {
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/agent/manager/manager_test.go:1187]: :+1:

File: pkg/agent/manager/manager_test.go [hunk 3]
```diff
@@ -1198,19 +1184,24 @@ func createCA(t *testing.T, clk clock.Clock) (*x509.Certificate, *ecdsa.PrivateK
...
+	svidKey, err := keymanager.ForSVID(km).GenerateKey(context.Background(), nil)
+	require.NoError(t, err)
+
+	return createSVIDWithKey(t, clk, ca, cakey, spiffeID, ttl, svidKey), svidKey
+}
+
+func createSVIDWithKey(t *testing.T, clk clock.Clock, ca *x509.Certificate, cakey *ecdsa.PrivateKey, spiffeID spiffeid.ID, ttl time.Duration, svidKey crypto.Signer) []*x509.Certificate {
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/agent/manager/manager_test.go:1194]: :+1:

File: pkg/agent/plugin/keymanager/base/keymanagerbase.go [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+		if opts.HashAlgorithm == keymanagerv1.HashAlgorithm_UNSPECIFIED_HASH_ALGORITHM {
+			return nil, status.Error(codes.InvalidArgument, "hash algorithm is required")
+		}
+		signerOpts = crypto.Hash(opts.HashAlgorithm)
+	case *keymanagerv1.SignDataRequest_PssOptions:
+		if opts.PssOptions == nil {
+			return nil, status.Error(codes.InvalidArgument, "PSS options are nil")
+		}
+		if opts.PssOptions.HashAlgorithm == keymanagerv1.HashAlgorithm_UNSPECIFIED_HASH_ALGORITHM {
+			return nil, status.Error(codes.InvalidArgument, "hash algorithm is required")
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/agent/plugin/keymanager/base/keymanagerbase.go:171]: :+1: (fixed in the server as well)

File: pkg/agent/plugin/keymanager/base/keymanagerbase.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/agent/plugin/keymanager/base/keymanagerbase.go:93]: :+1: (fixed in server as well)

File: pkg/agent/plugin/keymanager/base/keymanagerbase.go [hunk 5]
```diff
@@ -0,0 +1,338 @@
...
+		return nil, status.Errorf(codes.Internal, "unable to make key entry: %v", err)
+	}
+
+	return entry, nil
+}
+
+func makeKeyEntry(keyID string, keyType keymanagerv1.KeyType, privateKey crypto.Signer) (*KeyEntry, error) {
+	pkixData, err := x509.MarshalPKIXPublicKey(privateKey.Public())
+	if err != nil {
+		return nil, err
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/agent/plugin/keymanager/base/keymanagerbase.go:235]: :+1: (fixed on server as well)

File: pkg/agent/plugin/keymanager/disk/disk.go [hunk 6]
```diff
@@ -2,144 +2,244 @@ package disk
...
+	if err != nil {
+		return err
 	}
-	keyPath := path.Join(d.dir, keyFileName)
 
-	if err := diskutil.AtomicWriteFile(keyPath, req.PrivateKey, 0600); err != nil {
-		return nil, err
+	// Load the key from the deprecated key path. ONLY add this key to the
+	// entries list if there is not an entry for it already.
+	// TODO: stop doing this in 1.1
```
[the remark this answers is not in the document]
Comment (azdagron) [reply pkg/agent/plugin/keymanager/disk/disk.go:97]: Yes, for sure we should. I'll start a list.

File: pkg/agent/plugin/keymanager/base/keymanagerbase.go (hunk 1 above)
Comment (azdagron) [reply pkg/agent/plugin/keymanager/base/keymanagerbase.go:93]: yes, for sure. I've added the ones you've pointed out and added some code to wrap the error generically with a prefix like "failed to generate key"

**Review by amartinezfayo — APPROVED** [review]
:shipit:

**Pull request merged**